### ISSUE TITLE
user TA got read permission fault when 

### ISSUE THREAD

**Issue by AllenSiao** [OP-TEE/optee_os#1763]
Hi, 

I got read permission fault when I access the memory which malloc from CA. The log as below:

DEBUG:   [0x0@0x1] TEE-CORE:abort_handler:555: [abort] abort in User mode (TA will panic)
ERROR:   [0x0@0x1] TEE-CORE: 
ERROR:   [0x0@0x1] TEE-CORE: user TA prefetch-abort at address 0x272b200 (read permission fault)
ERROR:   [0x0@0x1] TEE-CORE:  fsr 0x0000000d  ttbr0 0x1627046a  ttbr1 0x1626c06a  cidr 0x1
ERROR:   [0x0@0x1] TEE-CORE:  cpu #1          cpsr 0x00000010
ERROR:   [0x0@0x1] TEE-CORE:  r0 0x00000000      r4 0x0271f140    r8 0x16276f40   r12 0x16276f94
ERROR:   [0x0@0x1] TEE-CORE:  r1 0x00102500      r5 0x02723180    r9 0x160190e5    sp 0x001026b0
ERROR:   [0x0@0x1] TEE-CORE:  r2 0x0000004b      r6 0x00102520   r10 0x00000000    lr 0x0027ab71
ERROR:   [0x0@0x1] TEE-CORE:  r3 0x00000000      r7 0x027271c0   r11 0x00000000    pc 0x0272b200
ERROR:   [0x0@0x1] TEE-CORE: Status of TA 2c5c47c9-eb7e-4c04-9aa8f787e759f23e (0x161906d8) (active)
ERROR:   [0x0@0x1] TEE-CORE: - load addr : 0x200000    ctx-idr: 1
ERROR:   [0x0@0x1] TEE-CORE: - stack: 0x100000 10240
ERROR:   [0x0@0x1] TEE-CORE: sect 0 : 0x100000 1048576
ERROR:   [0x0@0x1] TEE-CORE: sect 1 : 0x200000 1048576
ERROR:   [0x0@0x1] TEE-CORE: sect 2 : 0x300000 2097152
ERROR:   [0x0@0x1] TEE-CORE: sect 3 : 0 0
ERROR:   [0x0@0x1] TEE-CORE: sect 4 : 0x2700000 1048576
ERROR:   [0x0@0x1] TEE-CORE: sect 5 : 0x500000 16777216
ERROR:   [0x0@0x1] TEE-CORE: sect 6 : 0x1600000 16777216
ERROR:   [0x0@0x1] TEE-CORE: sect 7 : 0 0
ERROR:   [0x0@0x1] TEE-CORE: sect 8 : 0 0
ERROR:   [0x0@0x1] TEE-CORE: sect 9 : 0 0
ERROR:   [0x0@0x1] TEE-CORE: sect 10 : 0 0
DEBUG:   [0x0@0x1] TEE-CORE:user_ta_enter:1033: tee_user_ta_enter: TA panicked with code 0xdeadbeef

I am sure the address:0x272b200 is a valid address. So I don't know what happened about this issue.
I set the memory attribute as TEE_PARAM_TYPE_MEMREF_INOUT. 
Can anyone give me some advise?

BRs,
AllenSiao

**Comment by etienne-lms** [comment]
`prefetch-abort` looks like an attempt to execute from memory: here the memory is shared memory and hence not executable in the TA mapping.

**Comment by AllenSiao** [comment]
@etienne-lms
some question as below: 
1. As my understanding, the share memory go through TEE_Param will be mapping automatically by optee-os. So why TA mapping failed? 

2. I don't understand what why the address was be attempt to execute, it is just an address of share memory. can you explain it more accuracy? 

3. My request is that CA will malloc 2 continuously memory block separately, and these 2 memory would be split into several groups. Then TA get the address of all groups and handle it. And CA free the share memory when TA finish. Is my request possible?

**Comment by etienne-lms** [comment]
1. The TEE buffer params (memref) are mapped inside the TA but with read/write/no-exec permissions.

2. Check this from your TA implementation.
To get some more friendly info on where TA crashed, you can build optee_os with `CFG_TEE_CORE_DEBUG=y` (implies `CFG_UNWIND=y`). Once the you get the trace, you can use **scripts/symbolize.py** to get a human readable backtrace (thanks jerome:).

3. For the TA to access shared memory data, the CA must invoke the TA with the expected buffers passed as memref parameters at each TA invocation. There is no other way for a TA to access shared memory.
Note even if the same buffers are passed to the TA, the TA might get different virtual address for the buffer at each command invocation.

**Comment by AllenSiao** [comment]
@etienne-lms 
1. I expect TA just needs to read/write share memory but no executing. Therefore I am confusing why TA try to execute the address, and curious that why TA without permission to read the address?

3.  IF CA malloc a share memory in first invoke, then free the share memory in last invoke. Will the share memory exists in whole TA process?

**Comment by AllenSiao** [comment]
@etienne-lms 
I found that I trash myself then cause the read permission fault. Thanks for your answer anyway.

**Comment by etienne-lms** [comment]
> 2. IF CA malloc a share memory in first invoke, then free the share memory in last invoke. Will the share memory exists in whole TA process?

No. The fact is that a TA can access only its private memory (code/data/heap/stack) and the memory buffers passed as invocation parameter (memref). Hence, if CA allocates a shared memory buffer and invoke the TA with the buffer as param, the TA will access it. If the CA invokes again the TA without passing the buffer as param, the buffer is still allocated, but the TA cannot access it. So CA has to invoke the TA passing the buffer for every invocation.

**Comment by AllenSiao** [comment]
will the share memory reset after invoke finish? In other words, will the data in the share memory be cleaned?

**Comment by etienne-lms** [comment]
As long as the CA does not release a buffer allocated in the shared memory, only the CA can access/modifiy it, and eventually the TA if the CA invokes the TA with the buffer (or part of it) as a memref parameter.
